ENH: Boundary value problem solver for ODE systems

nmayorov opened the pull request:
Hi! 

This is my attempt to provide a Python implementation of a BVP solver. My main source was the paper "A BVP Solver Based on Residual Control and the MATLAB PSE".

`solve_bvp` implements a collocation algorithm for a cubic C^1 continuous spline with residuals (defect) control. It can solve problems with non-separated boundary conditions and unknown parameters (like eigenvalue-eigenfunction problems). I used my experience with trust-region methods (from implementing `least_squares`) to make an adaptation for a collocation system solver. 

My feeling is that the algorithm is reasonable, but perhaps some improvements can be discovered and implemented.

This [notebook](http://nbviewer.jupyter.org/urls/gist.githubusercontent.com/nmayorov/f8af5ca956c6a7f75ecdb578a2655894/raw/a4ae7e8edb3b1fb0cbb6d60788e84fcc6975cce1/solve_bvp_demo.ipynb) has examples of usage.

Now the list of potential TODO things:

- [x] Support for problems with a singular term. This [paper] (http://cs.smu.ca/~muir/JNAIAM_Shampine_Muir_Xu2006.pdf) on BVP_SOLVER contains a good section about that. I think it's at least a fun addition and can also be useful.
- [x] Allow user supplied Jacobians for a system RHS and boundary conditions. It's theoretically a useful thing and not difficult to add. But I'm not sure about its practicality. Forming a 3-dimensional stack of Jacobians for a ODE might be awkward for users. Also there are boundary conditions and unknown parameters, it can get a bit too complicated to specify all this. If most of the practical problems can be solved with finite difference approximation, maybe we don't need it. I guess we can decide here if it's necessary.
- [x] Try to solve more examples from the literature and see if any problems will occur.

Any feedback and code reviews are welcome. Note that the code is not very polished yet, consider it as a beta version.

nmayorov commented:
In 2 builds: https://travis-ci.org/scipy/scipy/jobs/120447558 and https://travis-ci.org/scipy/scipy/jobs/120447561 I got slightly worse accuracy in one test. 

Any ideas why it can happen? Should I be worried about that or simply increase tolerance to make a test pass? 

I assume that it could happen due to small changes in BLAS/LAPACK implementation (I use SVD and QR).

bjodah commented:
I think this looks really interesting. FWIW, I think allowing users to specify an explicit jacobian may be very valuable. If the derivation is tricky, at least I just resort to deriving it symbolically using SymPy and then create a callback using `sympy.lambdify` (when I am solving IVPs that is).

nmayorov commented:
@bjodah, noted your preference for analytical Jacobian.

I will come back to this PR soon. @pv your review and opinion is necessary.

nmayorov commented:
Now problems with the singular term can be solved. [Examples](http://nbviewer.jupyter.org/gist/nmayorov/f8af5ca956c6a7f75ecdb578a2655894) updated.

Next I will implement analytic Jacobian support and test on several other problems. After that I believe it will be ready. I'm not going to consider multi-point BVPs for now.

Sill looking for feedback.

nmayorov commented:
Two significant changes.

1. After discovering a problem where the original solver had problems, I decided to try a simple Newton method with backtracking (SuperLU is used to solve a sparse equation) and affine-invariant cost function. It was able to solve this problem better --- mostly by accident, because the problem had a bad starting guess and was actually quite hard. There is no clear advantage and the trust-region solver can potentially work better in other situations. But I decided to switch to the Newton method, because it's simpler, faster,  works quite well and conventional for BVPs so I could rely on the literature.

2. Added support for analytic Jacobians. You can pass function Jacobian, boundary condition Jacobian or both or neither.

Check out updated [examples](http://nbviewer.jupyter.org/gist/nmayorov/f8af5ca956c6a7f75ecdb578a2655894).

nmayorov commented:
I did a more comprehensive [benchmark](http://nbviewer.jupyter.org/urls/gist.githubusercontent.com/nmayorov/b9ba2add85c5cbc2ca08afa43131b1c0/raw/c566bad0444a93a906f1d1e1b1da9e3ef1e9e397/bvp_benchmarks.ipynb). It is very utilitarian and not very pleasant to read, but I believe it proved that the solver is ready for usage.

So now I check the last item and remove [WIP] tag.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,924 @@
+
+    dbc_dya = np.empty((n, n + k), dtype=dtype)
+    h = EPS**0.5 * (1 + np.abs(ya))
+    for i in range(n):
```
Small helper function could reduce repetition here.

nmayorov replied:
I thought about that at some point, because `estimate_fun_jac` also does the similar thing. But there are different shape of arrays and different calling signatures, so probably it will require the same amount of code and won't be very clean. I would prefer to keep this lame repetition style here.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,924 @@
+        values = np.hstack((values, dPhi_dp.ravel(), dbc_dp.ravel()))
+
+    J = coo_matrix((values, (i_jac, j_jac)))
+    return csc_matrix(J)
```
coo_matrix(..., copy=False), J.tocsc(copy=False) ?

nmayorov replied:
According to the signature `copy=False` by default (for both `coo_matrix` and `csc_matrix`). Is it fine then?

pv replied:
ok, indeed.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,924 @@
+                         df_dp_middle, dbc_dya, dbc_dyb, dbc_dp):
+    """Construct the Jacobian of the whole collocation system.
+
+    Returns
```
I think this could use a more explicit comment on what it is doing, so that it will be more easily understandable:

- what function exactly is differentiated (collocation_func)
- block structure of jacobian
- the formulas you wrote down in order to write this function (e.g. indicating what is the h^2 term)

nmayorov commented:
@pv, thanks for starting the review. I tried to explain `construct_global_jac` better. Check if it is more clear now.

pv commented:
I ported the test problem set from scikits.bvp1lg to solve_bvp, maybe useful: https://github.com/pv/scipy-work/blob/bvp/scipy/integrate/tests/test_bvp_problems.py
(A bit clunky as there are many converter layers...)

Some observations:

- `solve_bvp` can solve all problems except problem 7
- Averaged over all tests, the runtime is around factor of 4 larger than colnew. However, this is not fully comparable number since I'm not sure the tolerance is the same. Not bad!
- One issue with Problem7 is divergent term `~ 1/(x-a)^2` but colnew/colsys can apparently deal with it without knowing there is a singularity. Maybe it avoids evaluation of `f` at x=a,b and it works by luck? If so, perhaps solve_bvp could also attempt to work also in this case?
- A variant of Problem7 with `x -> x + 1` to remove the divergence fails to solve unless I increase the eps parameter 100x. Didn't look into why in more detail yet...
- `test_continuation` on Problem3 needed some tweaks to pass through what I think is a branching point. Probably nothing to worry, since the previous approach was tuned to colnew.

So performance-wise, it's looking fairly good to me based on testing so far.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,998 @@
+    dPhi_dy_0 = np.empty((m - 1, n, n), dtype=dtype)
+    dPhi_dy_0[:] = -np.identity(n)
+    dPhi_dy_0 -= h / 6 * (df_dy[:-1] + 2 * df_dy_middle)
+    T = np.einsum('...ij,...jk->...ik', df_dy_middle, df_dy[:-1])
```
Something like this will probably be faster than einsum:
```
def _stacked_matmul(a, b):
    """
    Stacked matrix multiply; out[i,:,:] = np.dot(a[i,:,:], b[i,:,:])
    """
    assert a.ndim == 3
    assert b.ndim == 3

    if a.shape[1] > 50:
        # BLAS gives speedups
        out = np.empty_like(a)
        for j in range(a.shape[0]):
            out[j] = np.dot(a[j], b[j])
        return out
    else:
        return np.einsum('...ij,...jk->...ik', a, b)
```
The speed difference is noticeable for problem10 in the scikits.bvp1lg problem set.

(Interestingly, recent numpy has `np.matmul` for this, but it's not faster than einsum :(

nmayorov replied:
Do you really recommend to do so? I didn't consider speed issues, so don't have any feel about it.

pv replied:
I would perhaps suggest to do it, since this seems to be a performace bottleneck if you have large numbers of variables.

nmayorov replied:
In Problem 10 I read
```
def f(self, x, u):
     return np.zeros((self.n_a, x.size))
```
Shouldn't we consider that `y0' = y1, y2' = y3, etc.`? Setting `f` to all-zeros doesn't seem right.

As for the speed-up. I tested on mesh with 500 nodes and 256 variables, and the time is still strongly in favor of Python loop + BLAS. So doing fast matrix multiplications is indeed more important that avoiding a Python loop. Good that you noticed it, because I probably would never think about that.

I believe (and did some rough tests) that the threshold of 50 is about right for the typical number of mesh nodes (like 50-500).

pv replied:
I think the FirstOrderConverter wrapper sets the 2nd derivative conditions. (colnew can take variable order equations in, but the mus solver couldn't, so the complexity dates back from that.)

nmayorov replied:
I see then.

nmayorov commented:
Great benchmark, especially time considerations.

> One issue with Problem7 is divergent term ~ 1/(x-a)^2 but colnew/colsys can apparently deal with it without knowing there is a singularity. Maybe it avoids evaluation of f at x=a,b and it works by luck? If so, perhaps solve_bvp could also attempt to work also in this case?

Indeed I read that some solvers (probably incl. COLSYS) work without evaluating derivatives at end points. And it can fail on some occasions.

This algorithm can't deal with such singularities (except for "linear singularities"). Actually BVP_SOLVER can't too, this can be considered as a disadvantage.

> A variant of Problem7 with x -> x + 1 to remove the divergence fails to solve unless I increase the eps parameter 100x. Didn't look into why in more detail yet...

It can be changed drastically by such modification (you sure know that too). One way to deal with such problems is to analyze the solution behavior near singularity, write a series and solve a problem shifting the boundary slightly from the singular point (and changing BC according to the series). By doing so an unknown parameter is introduced (usually derivative at the x=a). 

By no means I'm mastered this approach and not sure if I can do it for this problem, not sure if it's that necessary either. I'll check if it's possible to reformulate the problem like ``u = x phi, v = x psi``.

Other than that, I'm quite happy with the results.

pv commented:
`x->x+1` removes the divergence. The problem is of course then not related to the original problem any more --- however, colnew manages to solve the modified problem, but solve_bvp does not. There seem to be quite rapid changes on the edges, in a range determined by the `eps` parameter. For large values of `eps` the solutions agree with colnew, but solve_bvp stops being able to solve for `eps <~ 0.05`.

nmayorov commented:
It removes divergence, but I don't understand how the new problem is related to the original. But I got your point, that you still can compare the results.

The reformulation approach worked, and I was able to reproduce the final solution using continuation technique: 

https://gist.github.com/nmayorov/16aff3f7bd0eb67729d6c502bd4742b3

From the paper I don't understand what solution they get starting from (0, 0) guess. I got this weird solution with a "sawtooth", but it requires a steady continuation. Does COLSYS really converges to it starting from (0, 0) guess? If so, it's really impressive and sort of surprising.

pv commented:
Yes, colnew converges to that solution from zero guess. This is probably sensitive to how exactly the nonlinear solver is done.

The modified problem is not related to the old one (I "found" it, as I wanted to get a non-singular problem to check if the functions were returning output in correct format). However, it could be a potentially useful test problem, as the boundary layers seem to cause some problems for solve_bvp. Of course, there probably are test problems with similar features and analytical solutions available.

pv commented:
I see some papers point out the colsys approach to edge divergences is not robust, may give false solutions. So probably the approach here (force users to reparameterize, except for 1/x terms) is reasonable.

nmayorov commented:
> However, it could be a potentially useful test problem, as the boundary layers seem to cause some problems for solve_bvp. Of course, there probably are other test problems with similar features and analytical solutions available.

I would say that the problem with eps=1e-3, mu=1e-3 is extremely stiff (if it's the correct term), because it has very small 1e-9 coefficient for the leading derivative. I think it's natural and expected for numerical solvers to experience difficulties for such problems. 

Maybe COLSYS is better than other solvers in this regard. But from a quick glance their approach might be substantially more difficult (it involves B-splines for starters).

There is another test collection with analytic solutions for some problem http://wwwf.imperial.ac.uk/~jcash/BVP_software/PROBLEMS.PDF Would it be good / necessary to make another test suite? No "standard" initial guess provided, so I'm not sure how exactly to implement this benchmark.

> I see some papers point out the colsys approach to edge divergences is not robust, may give false solutions. So probably the approach here (force users to reparameterize, except for 1/x terms) is reasonable.

I noticed that too. Anyway I don't think it's very easy or practical to adopt COLSYS approach now.

nmayorov commented:
@pv, what is the current state then? Any specific suggestions or wishes?

pv commented:
Ok, so I think the code is a clear and short implementation, and if you have the papers at hand it's possible to understand all parts, in case someone else needs to dig into this later on. I think the testing so far indicates it's probably correct.

What I'm not yet fully sure is how this stacks against the "old" solvers (colnew, twpbvp, bvp_solver) with regard to number of mesh nodes and other metrics. For colnew, we saw runtime is quite similar and I think the mesh usage was similar (maybe larger), but I didn't check how the accuracy should be compared. J. Cash has a [table of results for twpbvp](http://wwwf.imperial.ac.uk/~jcash/BVP_software/twpbvp/twpbvpresults.php) for his 35 test problems, some with boundary layers. Comparing some selected ones could be informative.

In the test problems in `test_bvp.py`, I'd like to have some hardcoded check for the number of mesh points, `assert_(len(sol.x) < some_number)` so that in case the implementation is changed later things don't get worse unnoticed. Perhaps adding a test problem with a boundary layer could be useful to ensure we don't break the mesh selection later on.

Not tested: the `> 50` branch in `stacked_matmul` (maybe add a big test problem or test directly). Singular jacobians. Printing (maybe just use `scipy._lib._testutils.suppressed_stdout` and run something with verbose=2).

No performance benchmark: adding a small timing benchmark could be useful in keeping track of performance, but not necessary before merging.

I'm not fully sure if the complex analytic restriction will cause confusion. Maybe I've read too much stackoverflow, so let's leave it in for now, I think it's useful --- in the other alternative of z->[re,im] mapping, the jacobians become bit messy to specify, so it's clearer to leave that to the user.

Apart from those (plus some small comment to follow), +1 from me for merging.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,1013 @@
+        Right-hand side of the system. The calling signature is
+        ``fun(x, y, p)``, where ``x`` is ndarray of shape (m,) and ``y`` is
+        ndarray of shape (n, m), meaning that ``y[:, i]`` corresponds to
+        ``x[i]``, ``p`` is optional with shape (k,). The return value must
```
Saying "p is optional" is a bit ambigous. This could be clearer: "The calling signature is `fun(x, y)` or `fun(x, y, p)` if parameters are present. Here, x is ..."

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,1013 @@
+
+    Examples
+    --------
+    We solve a simple Sturm-Liouville problem::
```
It could also be useful to a second example without unknown parameters.

pv commented on scipy/integrate/_bvp.py:
```diff
@@ -0,0 +1,1013 @@
+    Problems can be solved in a complex domain as well. In this case y and p
+    are considered to be complex, and f and bc are assumed to be complex-valued
+    functions, but x stays real. Note that f and bc must be complex
+    differentiable (satisfy Cauchy–Riemann conditions) in this case, otherwise
```
The - in "Cauchy-Riemann" is a non-ASCII character. Maybe better to replace with the ascii "-", as it seems to break Sphinx. (Or add `# -*- encoding:utf-8 -*-` on top of the file.)

nmayorov commented:
I addressed all code issues. Didn't include tests for printing though, but I tested all cases deliberately in jupyter (I think it's more reasonable for this kind of things).

> What I'm not yet fully sure is how this stacks against the "old" solvers (colnew, twpbvp, bvp_solver) with regard to number of mesh nodes and other metrics. For colnew, we saw runtime is quite similar and I think the mesh usage was similar (maybe larger), but I didn't check how the accuracy should be compared. J. Cash has a table of results for twpbvp for his 35 test problems, some with boundary layers. Comparing some selected ones could be informative.

I compared results with bvp4c with the equivalent tolerance settings: numbers of nodes were the same or slightly higher, because bvp4c can delete nodes too (but usually only couple of them, doesn't look significant).

So I'll pick some problems from J. Cash collection and check the results in a "manual mode" (to adjust tolerances).

pv commented:
Here's the print test: https://github.com/pv/scipy-work/commit/d2b03270874868e
Testing in jupyter is OK, but it's better to automate.

pv commented:
I sent a mail to the list, I suggest we wait a couple of days in case someone has comments, but if not we can proceed.

nmayorov commented:
Thanks for the test code. 

After a few days I will be not very much online for some time. So it can stay here for people to review, it won't be disappointing for me. Or, if you prefer, we can try to finish it faster.

nmayorov commented:
Tested on some problems from J. Cash suite https://gist.github.com/nmayorov/437fb08d4d4ced99a8c6b95216d9566f 

I compute "error" as the mean (by nodes) relative difference between the solution and reference function. Most likely this is what is meant by "Error" in J. Cash tables (I checked the code and almost sure). I always use a zero initial guess and the mesh with 5 nodes.

My interpretation is that `solve_bvp` results are not terrible, maybe not as good as reported in the table (but I don't completely understand the methodology used to produce the table).

nmayorov commented:
Hey @pv. In view of no feedback from other people, what do you think we can or should do next?

pv commented:
I think if I still have some suggestions, but if not I think I will 
merge during the weekend.

nmayorov commented:
Cool, I will re-read the code.

nmayorov commented:
I was thinking about renaming `res` to `r` in the return bunch. My consideration is that "res" is often interpreted as "result", and I tend to name the return as `res`, so we have `res.res`, which I don't like. Another option is to remove `res` completely, I decided to include it mostly for the correctness monitoring.

This is very minor, maybe not necessary to change.

pv commented:
Thanks, it's in. We can revise later if needed.

Juanlu001 commented:
This is awesome, thanks all! :tada:
